### PR TITLE
Do not overwrite allowMSizeOptimization

### LINKED ISSUE #7098: [Yul] optimizer optimizes out side-effects that change the value returned by msize()?
## Description

Found by oss-fuzz: issue 15818

```
{
	let a

	function foo_singlereturn_0() -> out
	{
		mstore(lt(or(gt(1,or(or(gt(or(or(or(1,gt(or(gt(or(or(keccak256(pc(),or(gt(not(pc()),1),1)),1),not(1)),pc()),1),pc())),lt(or(1,sub(pc(),1)),1)),pc()),1),1),gt(not(pc()),1))),1),1),1)
		sstore(not(pc()),1)
	}

	function foo_singlereturn_1(in_1, in_2) -> out
	{
		extcodecopy(1,msize(),1,1)
	}

	a := foo_singlereturn_0()
	sstore(0,0)
	sstore(2,1)

	a := foo_singlereturn_1(calldataload(0),calldataload(3))
	sstore(0,0)
	sstore(3,1)
}
```

produces the following trace pre and post optimization

```
// Pre optimization trace
  EXTCODECOPY(1, 128, 1, 1)
Memory dump:
     0: 0000000000000000000000000000000000000000000000000000000000000001
    80: 6f00000000000000000000000000000000000000000000000000000000000000
Storage dump:
  0000000000000000000000000000000000000000000000000000000000000002: 0000000000000000000000000000000000000000000000000000000000000001
  0000000000000000000000000000000000000000000000000000000000000003: 0000000000000000000000000000000000000000000000000000000000000001
  ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff88: 0000000000000000000000000000000000000000000000000000000000000001
```

```
// Post optimization trace
  EXTCODECOPY(1, 32, 1, 1)
Memory dump:
     0: 0000000000000000000000000000000000000000000000000000000000000001
    20: 6f00000000000000000000000000000000000000000000000000000000000000
Storage dump:
  0000000000000000000000000000000000000000000000000000000000000002: 0000000000000000000000000000000000000000000000000000000000000001
  0000000000000000000000000000000000000000000000000000000000000003: 0000000000000000000000000000000000000000000000000000000000000001
  ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff88: 0000000000000000000000000000000000000000000000000000000000000001
```

## Environment

- Compiler version: latest develop
- Target EVM version (as per compiler settings): petersburg

## Steps to Reproduce

To obtain pre-optimization trace, do

```
$ yulrun --input-file <code.yul> run
Trace:
  EXTCODECOPY(1, 128, 1, 1)
Memory dump:
     0: 0000000000000000000000000000000000000000000000000000000000000001
    80: 6f00000000000000000000000000000000000000000000000000000000000000
Storage dump:
  0000000000000000000000000000000000000000000000000000000000000002: 0000000000000000000000000000000000000000000000000000000000000001
  0000000000000000000000000000000000000000000000000000000000000003: 0000000000000000000000000000000000000000000000000000000000000001
  ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff88: 0000000000000000000000000000000000000000000000000000000000000001
```

To obtain post-optimized trace, first obtain optimized yul code like so
```
$ solc --strict-assembly --optimize <code.yul> 2>/dev/null | | grep -Pzo "code {\n(.|\n)*}\n\n"
code {
        {
            let _1 := gt(not(pc()), 1)
            let _2 := pc()
            let _3 := not(0)
            let _4 := lt(or(1, add(pc(), _3)), 1)
            let _5 := pc()
            let _6 := pc()
            mstore(lt(or(gt(1, or(or(gt(or(or(or(gt(or(gt(_3, _6), 1), _5), _4), _2), 1), 1), _1), 1)), 1), 1), 1)
            sstore(not(pc()), 1)
            sstore(0, 0)
            sstore(2, 1)
            pop(foo_singlereturn_1(calldataload(0), calldataload(3)))
            sstore(0, 0)
            sstore(3, 1)
        }
        function foo_singlereturn_1(in, in_1) -> out
        { extcodecopy(1, msize(), 1, 1) }
    }
}
```
Then run yulrun against the optimized yul code
```
$ yulrun --input-file <optimized_yul_snippet_shown_above.yul> run
Trace:
  EXTCODECOPY(1, 32, 1, 1)
Memory dump:
     0: 0000000000000000000000000000000000000000000000000000000000000001
    20: 6f00000000000000000000000000000000000000000000000000000000000000
Storage dump:
  0000000000000000000000000000000000000000000000000000000000000002: 0000000000000000000000000000000000000000000000000000000000000001
  0000000000000000000000000000000000000000000000000000000000000003: 0000000000000000000000000000000000000000000000000000000000000001
  ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff88: 0000000000000000000000000000000000000000000000000000000000000001
```

### PR DESCRIPTION
**Pull request by chriseth**
Fixes https://github.com/ethereum/solidity/issues/7098

### CODE REVIEW

File: libyul/optimiser/UnusedPruner.cpp [hunk 1]
```diff
@@ -119,15 +120,13 @@ void UnusedPruner::operator()(Block& _block)
 void UnusedPruner::runUntilStabilised(
 	Dialect const& _dialect,
 	Block& _ast,
-	bool _allowMSizeOptization,
+	bool _allowMSizeOptimization,
 	set<YulString> const& _externallyUsedFunctions
 )
 {
-	_allowMSizeOptization = !SideEffectsCollector(_dialect, _ast).containsMSize();
```
Comment (chriseth) [inline libyul/optimiser/UnusedPruner.cpp:126]: This is the actual fix.

File: libyul/optimiser/UnusedPruner.cpp [hunk 2]
```diff
@@ -26,6 +26,7 @@
 #include <libyul/Exceptions.h>
 #include <libyul/AsmData.h>
 #include <libyul/Dialect.h>
+#include <libyul/AsmPrinter.h>
```
Comment (bshastry) [inline libyul/optimiser/UnusedPruner.cpp]: Is this header required?

File: libyul/optimiser/UnusedPruner.cpp (hunk 1 above)
Comment (bshastry) [reply libyul/optimiser/UnusedPruner.cpp:126]: I understand that this was the bug. But was there any reasoning behind `_allowMsizeOptimization` being recomputed on the `_ast` and not trusting the value passed?

File: Changelog.md [hunk 3]
```diff
@@ -12,6 +12,7 @@ Compiler Features:
  * Metadata: Update the swarm hash, changes ``bzzr0`` to ``bzzr1`` and urls to use ``bzz-raw://``.
  * Standard JSON Interface: Compile only selected sources and contracts.
  * Standard JSON Interface: Provide secondary error locations (e.g. the source position of other conflicting declarations).
+ * Yul Optimizer: Fix an issue where memory-accessing code was removed even though ``msize`` was used in the program.
```
Comment (bshastry) [inline Changelog.md]: Shouldn't this move to `Bugfixes:` section?

**Comment by chriseth** [comment]
Updated.

File: test/libyul/yulOptimizerTests/stackCompressor/unusedPrunerWithMSize.yul [hunk 4]
```diff
@@ -0,0 +1,36 @@
+{
+    {
+        let _13_71 := 1
+        let _17_72 := pc()
+        let _22_75 := pc()
+        let _25_76 := pc()
+        let _30_80 := pc()
+        let _32_81 := pc()
+        // This should not be removed
+        pop(keccak256(1, 2))
```
Comment (bshastry) [inline test/libyul/yulOptimizerTests/stackCompressor/unusedPrunerWithMSize.yul:10]: This test case passes on both unpatched and patched versions of the fuzzing binary (tested on https://github.com/ethereum/solidity/blob/develop/test/tools/ossfuzz/yulProto_diff_ossfuzz.cpp). 

The original test case correctly fails on the unpatched version and passes once this PR is applied. Please consider using the original test or adapting it such that we have a fail/pass transition pre/post PR.

**Review by bshastry — CHANGES_REQUESTED** [review]
LGTM apart from the test input not making the unpatched version fail.

File: libyul/optimiser/UnusedPruner.cpp (hunk 1 above)
Comment (chriseth) [reply libyul/optimiser/UnusedPruner.cpp:126]: I think it was just a copy-paste error from the function that does not have the parameter.

File: test/libyul/yulOptimizerTests/stackCompressor/unusedPrunerWithMSize.yul (hunk 4 above)
Comment (chriseth) [reply test/libyul/yulOptimizerTests/stackCompressor/unusedPrunerWithMSize.yul:10]: It does not pass for me. Are you sure?

Comment (chriseth) [reply test/libyul/yulOptimizerTests/stackCompressor/unusedPrunerWithMSize.yul:10]: This actually proves that the test is correct: https://circleci.com/gh/ethereum/solidity/150576?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

**Pull request merged**